vms/platformvm/network/warp.go

```diff
@@ -0,0 +1,358 @@
+		}
+	}
+
+	return nil
+}
+
+func (s signatureRequestVerifier) verifySubnetValidatorRegistration(
+	msg *message.SubnetValidatorRegistration,
+	justificationBytes []byte,
+) *common.AppError {
+	if msg.Registered {
+		return s.verifySubnetValidatorRegistered(msg.ValidationID)
+	}
+
+	var justification platformvm.SubnetValidatorRegistrationJustification
+	if err := proto.Unmarshal(justificationBytes, &justification); err != nil {
+		return &common.AppError{
+			Code:    ErrFailedToParseJustification,
+			Message: "failed to parse justification: " + err.Error(),
+		}
+	}
+
+	switch preimage := justification.GetPreimage().(type) {
+	case *platformvm.SubnetValidatorRegistrationJustification_ConvertSubnetTxData:
+		return s.verifySubnetValidatorNotCurrentlyRegistered(msg.ValidationID, preimage.ConvertSubnetTxData)
+	case *platformvm.SubnetValidatorRegistrationJustification_RegisterSubnetValidatorMessage:
+		return s.verifySubnetValidatorCanNotValidate(msg.ValidationID, preimage.RegisterSubnetValidatorMessage)
+	default:
+		return &common.AppError{
+			Code:    ErrInvalidJustificationType,
+			Message: fmt.Sprintf("invalid justification type: %T", justification.Preimage),
+		}
+	}
+}
+
+// verifySubnetValidatorRegistered verifies that the validationID is currently a
+// validator.
+func (s signatureRequestVerifier) verifySubnetValidatorRegistered(
+	validationID ids.ID,
+) *common.AppError {
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	// Verify that the validator exists
+	_, err := s.state.GetSubnetOnlyValidator(validationID)
+	if err == database.ErrNotFound {
+		return &common.AppError{
+			Code:    ErrValidationDoesNotExist,
+			Message: fmt.Sprintf("validation %q does not exist", validationID),
+		}
+	}
+	if err != nil {
+		return &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: "failed to get subnet only validator: " + err.Error(),
+		}
+	}
+	return nil
+}
+
+// verifySubnetValidatorNotCurrentlyRegistered verifies that the validationID
+// could only correspond to a validator from a ConvertSubnetTx and that it is
+// not currently a validator.
+func (s signatureRequestVerifier) verifySubnetValidatorNotCurrentlyRegistered(
+	validationID ids.ID,
+	justification *platformvm.SubnetIDIndex,
+) *common.AppError {
+	subnetID, err := ids.ToID(justification.GetSubnetId())
+	if err != nil {
+		return &common.AppError{
+			Code:    ErrFailedToParseSubnetID,
+			Message: "failed to parse subnetID: " + err.Error(),
+		}
+	}
+
+	justificationID := subnetID.Append(justification.GetIndex())
+	if validationID != justificationID {
+		return &common.AppError{
+			Code:    ErrMismatchedValidationID,
+			Message: fmt.Sprintf("validationID %q != justificationID %q", validationID, justificationID),
+		}
+	}
+
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	// Verify that the provided validationID either:
+	// - Is in the current state
+	// - Was removed from the current state
+	// - Was not included in the subnet conversion
```

Comment:
At first pass, this comment confused me because I thought we should be verifying the the validationID is **not** in the current state.

```suggestion
	// Verify that the provided subnetID has been converted 
	// and that the provided validationID is not in the current state.
	// This means the validationID was either not included in the subnet
	// conversion, or that it was removed from the current state.
```